File: dart/lib/src/events.dart
// The solo event surface (design station.md 6): a bounded ring buffer
// plus a live tap with serialized callbacks. Events never fail an
// operation; overflow drops oldest and the drop count is visible in
// status().
//
// A port of typescript/src/events.ts, which is canonical. Events are
// plain string-keyed maps (the StationEvent v1 shape of station.md 6):
// { t, plugin?, corr?, kind, op?, http?, err?, meta? }.

typedef TapFn = void Function(Map<String, dynamic> ev);

class EventBuffer {
  final List<Map<String, dynamic>> _ring = [];
  final int _max;
  int _drops = 0;
  final List<TapFn> _taps = [];

  EventBuffer([int? max]) : _max = max ?? 1000;

  void emit(Map<String, dynamic> ev) {
    _ring.add(ev);
    if (_ring.length > _max) {
      _ring.removeAt(0);
      _drops++;
    }
    // Serialized, and a throwing tap must not fail the operation that
    // emitted the event.
    for (final fn in List<TapFn>.from(_taps)) {
      try {
        fn(ev);
      } catch (_e) {
        // A tap failure is the tap's problem, never the operation's.
      }
    }
  }

  List<Map<String, dynamic>> events() => List<Map<String, dynamic>>.from(_ring);

  void Function() tap(TapFn fn) {
    _taps.add(fn);
    return () {
      _taps.remove(fn);
    };
  }

  Map<String, dynamic> status() => <String, dynamic>{
        'buffered': _ring.length,
        'dropped': _drops,
      };
}

File: typescript/src/events.ts
import { StationEvent } from './types'

// The solo event surface (design §6): a bounded ring buffer plus a live
// tap with serialized callbacks. Events never fail an operation; overflow
// drops oldest and the drop count is visible in status().

export class EventBuffer {
  private ring: StationEvent[] = []
  private max: number
  private drops = 0
  private taps: ((ev: StationEvent) => void)[] = []

  constructor(max?: number) {
    this.max = null == max ? 1000 : max
  }

  emit(ev: StationEvent): void {
    this.ring.push(ev)
    if (this.ring.length > this.max) {
      this.ring.shift()
      this.drops++
    }
    // Serialized, and a throwing tap must not fail the operation that
    // emitted the event.
    for (const fn of this.taps) {
      try { fn(ev) } catch (_e) { }
    }
  }

  events(): StationEvent[] {
    return this.ring.slice()
  }

  tap(fn: (ev: StationEvent) => void): () => void {
    this.taps.push(fn)
    return () => {
      const i = this.taps.indexOf(fn)
      if (-1 !== i) { this.taps.splice(i, 1) }
    }
  }

  status(): { buffered: number, dropped: number } {
    return { buffered: this.ring.length, dropped: this.drops }
  }
}
